spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala

```diff
@@ -176,7 +176,12 @@ class IcebergSparkSqlExtensionsParser(delegate: ParserInterface) extends ParserI
   }
 
   private def isIcebergCommand(sqlText: String): Boolean = {
-    val normalized = sqlText.toLowerCase(Locale.ROOT).trim().replaceAll("\\s+", " ")
+    val normalized = sqlText.toLowerCase(Locale.ROOT).trim()
+      // Catch all SQL comments, including simple comments (that start with --), as well as multiline comments
+      .replaceAll("(?ms)/\\*.*?\\*/|--.*?\\n", " ")
+      // Replace any remaining newlines
+      .replaceAll("\\s+", " ")
+      .trim()
```

Comment:
Suggest to separate the logic of line 179-182 to a new method `normalizeSqlText()`, it is weird as a part of method `isIcebergCommand`. So that the caller side would be
```
 val sqlTextAfterSubstitution = substitutor.substitute(sqlText)
 val normalizedSqlText = normalizeSqlText(sqlTextAfterSubstitution)
    if (isIcebergCommand(sqlTextAfterSubstitution)) {
```

Reply:
This sounds reasonable to me too

Reply:
Assuming you mean to pass the result of `normalizeSqlText` to `isIcebergCommand`, my concern is that we want to allow the parser to do the actual removal of comments.

I’m happy to move the normalization into a separate function, but I don’t think we should be manipulating the SQL text beyond whether or not we are checking if this is Iceberg specific DDL.

If the request is to just make it a function that normalizes and the normalized result is _only_ passed to `isIcebergCommand`, I’m not opposed to that. Though I do think it opens up the possibility of accidentally refactoring and then continuing to pass the stripped out SQL text to the parser.

The parser is responsible for removing comments etc, as it knows how to differentiate between comments that are in lined SQL hints etc.

I could make a named function that is then called inside of `isIcebergCommand` though (or whose results are _only_ passed to `isIcebergCommand`).

Reply:
Ok. I updated it to use a named function, but the constraints on that function are pretty large as we want the antlr parser to be responsible for removing SQL comments etc.

I think it's better the original way and that the best solution would be to add additional comments clarifying why we're manipulating the `sqlText` within `isIcebergCommand`. I do concede that overall it is somewhat weird to be normalizing the SQL text directly within that function, but doing so helps prevent people from accidentally calling `normalizeSqlText` on sql text that will be passed to the parser.

So my suggestion would be something more like:
```scala
  private def isIcebergCommand(sqlText: String): Boolean = {
    // The sqlText has not been passed through the parser yet and therefore has not had
    // unimportant comments and unnecessary whitespace and newlines removed.
    // 
    // SQL comments and newlines are removed here so that Iceberg commands can be properly recognized,
    // while still allowing the antlr parser to perform the actual work of differentiating
    // comments that don't have any meaning to the query from comments that behave as sql hints etc.
    val normalized = sqlText.toLowerCase(Locale.ROOT).trim()
      // Catch all SQL comments, including simple comments (that start with --), as well as comments
      // of the form /* ... */, including those that span multiple lines.
      .replaceAll("(?ms)/\\*.*?\\*/|--.*?\\n", " ")
      // Replace any remaining newlines
      .replaceAll("\\s+", " ")
      .trim()
    normalized.startsWith("call") || (
        normalized.startsWith("alter table") && (
            normalized.contains("add partition field") ||
            normalized.contains("drop partition field") ||
            normalized.contains("replace partition field") ||
            normalized.contains("write ordered by") ||
            normalized.contains("write locally ordered by") ||
            normalized.contains("write distributed by") ||
            normalized.contains("write unordered") ||
            normalized.contains("set identifier fields") ||
            normalized.contains("drop identifier fields")))
  }
```

I'm not in love with this either, but I think what's making things a bit weird is the fact that we have to check whether or not something is an Iceberg specific command before using the actual parser.

I'm not prepared to refactor the parser to that level yet, so I think his is the best solution. That said, if we want to use the current `normalizeSqlText` function with the comment / understanding that it should only be used in that one spot, that is a decision I would leave up to the community. I personally think it's unsafe in the long term to have a function, `normalizeSqlText`, that is only meant to be used in one very specific location and which could have dangerous side effects if used elsewhere.

Reply:
Thanks for the explain. Sorry, I miss that point. But I see another issue here. The method `isIcebergCommand` isn't reliable. For example, you can have a statement like this. 
```
alter table table_name SET TBLPROPERTIES ('key1' = 'add partition field')
```
It isn't an Iceberg command, but `isIcebergCommand` will return true since it just searches for the string `add partition field`.

One solution is to use the parser to get all tokens first, then decide if it is an Iceberg command.

Reply:
That's a fair assessment. However, I think that should be handed in a separate PR. As it's unrelated to this issue / is an existing issue. And admittedly I'm not going to have time to make large changes to the existing parsing code.

So I agree that there's definitely cases it doesn't handle well, but I wouldn't say that this makes it any worse and it does make it better for end users who would like to use procedures with DBT.

Reply:
Yeah, I agree that the table property issue is unrelated. I think it's fine either way. Given that we would only call normalizeSqlText in one place, I don't see a lot of value in separating it out just yet.

Reply:
Yeah it would be nice to do that as the parser already takes care of this, but I'm not quite sure how either.

I suspect that it's more work than it probably initially seems.